### DISCUSSION TITLE
Curious about duplicate functions.

### DISCUSSION THREAD

**Discussion by davidbruce** [JOML-CI/JOML#282] [General]
I've been a big fan of LWJGL and I have recently been exploring JOML.  
I'm curious why all the matrix classes have a public and a private functions for the same thing.  Is there some kind of performance benefit to this somewhere I am not seeing?  
e.g.:
https://github.com/JOML-CI/JOML/blob/7fd5da7d2c36c6a9d2383ab52815de020532b94e/src/org/joml/Matrix2f.java#L165-L168

https://github.com/JOML-CI/JOML/blob/7fd5da7d2c36c6a9d2383ab52815de020532b94e/src/org/joml/Matrix2f.java#L210-L213
 
Also, why have class fields and getters both be set as public, is it just for convenience?  Have abstract classes been considered to replace the interfaces to unify these, or is there some kind of performance penalty?  

Thank you for your time!

**Comment by httpdigest** [comment]
> I'm curious why all the matrix classes have a public and a private functions for the same thing

The two different "matrix element setter" methods are primarily used in the `Matrix4f` or `Matrix4x3f` classes, which know about "matrix properties" to accelerate matrix operations (like translate, rotate, scale, mul, ...) when we know about certain properties of the matrix.

Other classes, like the mentioned `Matrix2f` do not (yet) track/have such matrix properties, because it didn't pay off yet performance-wise. But for symmetry and consistency, these two flavors of matrix element setter methods are already in place.

> Have abstract classes been considered to replace the interfaces to unify these [...]

Indeed, that was the plan at some point and it might be reconsidered when there are actually two different feasible matrix/vector implementations when Panama's Vector API is stable and fast.

With that being said, please note that using the element setter methods or setting the element fields directly is really bad practice when using JOML (at least when it comes to the `4x4` and `4x3` classes), because of the mentioned matrix properties that are being read and written by the various transformation methods. When you set matrix elements explicitly (in the `4x4` and `4x3` classes) then you will essentially disable a lot of performance benefits because those methods will then conservatively throw away any matrix property information and always use the _generic_ matrix transformation methods which are far slower than their specializations. The matrix classes in JOML provide a multitude of matrix setup/creation methods, so when you find yourself explicitly setting matrix elements then you should look out for a method that already does this itself.
`Matrix4f.setPerspective()/perspective()` and `Matrix4f.setLookAt/lookAt()` are popular examples of methods that people simply write themselves and set the matrix elements explicitly, but that will disable the mentioned performance improvements.

**Reply by davidbruce** [reply]
Thanks for the info I appreciate it!  I guess I didn't get far enough down the class list to the 4x4 and 4x3 classes.

If the fields shouldn't be accessed what's the reason for keeping them public? Are there situations when you want to lose the properties?

**Reply by httpdigest** [reply]
> If the fields shouldn't be accessed what's the reason for keeping them public? Are there situations when you want to lose the properties?

On the 4x4 and 4x3 classes, the fields aren't public. And the 2x2 and 3x3 classes do not yet track properties, so there wasn't a need to make the fields non-public, yet.
